File: baekjoon/이분탐색,삼분탐색/10816/Main.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] card;
    static int N, M;

    public static void main(String[] args) throws Exception {
        System.setIn(new FileInputStream("C:\\Users\\evergreen\\Documents\\hello-spring\\untitled\\src\\input.txt"));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        card = new int[N];
        for (int i = 0; i < N; i++) {
            card[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(card); // nlogn
        M = Integer.parseInt(br.readLine());
        st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < M; i++) {
            int input = Integer.parseInt(st.nextToken());
            sb.append(binarySearch(input) + " ");
        }


        bw.write(sb.toString());
        bw.flush();
        bw.close();
    }

    /* 집 간 최소 거리는 1, 최대 거리는 가장 먼 거리*/
    static int binarySearch(int target) {
        int start = 0;
        int end = N;
        //lower bound -> target이 나타나는 첫 번째 인덱스를 반환
        while (start < end) {
            int mid = (start + end) / 2;
            if (card[mid] < target) {
                start = mid + 1;
            } else {
                end = mid;
            }
        }
        int lowerBound = start;
        if(card[lowerBound] != target) return 0;
        start = 0;
        end = N;
        //upper bound -> 중복 요소중 마지막 target 다음 값
        while (start < end) {
            int mid = (start + end) / 2;
            if (card[mid] <= target) {
                start = mid + 1;
            } else {
                end = mid;
            }
        }
        int upperBound = start;
        return upperBound - lowerBound;

    }

}

File: baekjoon/이분탐색,삼분탐색/10816/main.kt
fun main() {
    with(System.`in`.bufferedReader()) {
        val N = readLine().toInt()
        val cards = readLine().split(" ").map(String::toInt).sorted()
        val M = readLine().toInt()
        readLine().split(" ").map {
            cards.upperBound(it.toInt()) - cards.lowerBound(it.toInt())
        }.also {
            print(it.joinToString(" "))
        }

    }
}

fun List<Int>.lowerBound(target: Int): Int {
    var start = 0
    var end = size
    while (start < end) {
        val mid = (start + end) / 2
        if (target <= get(mid)) { // <=에서 target을 찾아도 범위를 줄여버리니까, else에서 start를 mid+1로 올려버리면 while문이 끝난다.
            end = mid
        } else {
            start = mid + 1
        }
    }
    return start
}

fun List<Int>.upperBound(target: Int): Int {
    var start = 0
    var end = size
    while (start < end) {
        val mid = (start + end) / 2
        if (target < get(mid)) { // 현재 값보다 타겟이 작으면, end를 mid로
            end = mid
        } else {
            start = mid +1
        }
    }
    return start
}

/*
* 이분탐색의 upper bound 와 lower bound 개념을 이용해서 중복된 원소가 나오는 첫번째 인덱스와, 마지막 인덱스를 구해야 한다.
*
* */